processor/processor.go

```diff
@@ -572,11 +560,23 @@ func (wp *workerPool) download(ctx context.Context, j *job.Job, validator *mimet
 		// This actually indicates a malformed url
 		return derrors.E("creating request", err)
 	}
-	if wp.p.UserAgent != "" {
+
+	if j.UserAgent != "" {
+		req.Header.Set("User-Agent", j.UserAgent)
+	} else if wp.p.UserAgent != "" {
 		req.Header.Set("User-Agent", wp.p.UserAgent)
 	}
 
-	resp, err := wp.p.Client.Do(req.WithContext(ctx))
+	// DownloadTimeout might be different than zero in case Job has been initalized
+	// with custom valid timeout, so this timeout will be used intead of the default.
+	if j.DownloadTimeout > 0 {
+		var cancel context.CancelFunc
+		timeout := time.Duration(j.DownloadTimeout)
+		ctx, cancel = context.WithTimeout(ctx, timeout*time.Second)
```

Comment:
If I get this correclty, this won't work as expected. By using a cancellation context we do not re-set the HTTP client timeout. This means we are capped by the timeout set in client, which is the default timeout [here](https://github.com/skroutz/downloader/pull/14/files#diff-1d1d00824b3268cdd91ecb964e33cddaR806).

By doing a simple test with your branch 

```
curl -d '{"download_timeout": 120, "aggr_id":"aggrFooBar", "aggr_limit":8, "url":"http://www.cartographicperspectives.org/index.php/journal/article/download/cp43-complete-issue/577", "callback_type":"kafka" ,"callback_dst":"dwl_images", "extra":"foobar"}' http://localhost:8000/download
```

It fails after 10 seconds

```
08:27:18 processor.1 | [processor] [worker pool:aggrFooBar] 2019/04/16 08:27:18 Started working...
08:27:19 api.1       | ts=2019-04-16T05:27:19.413229721Z app=downloader file=api.go lineno=232 function=(*API).ServeHTTP host=gallifrey component=api aggregation_id=aggrFooBar aggregation_limit=8 job_id=aiKAl8wnyoZhBw job_url=http://www.cartographicperspectives.org/index.php/journal/article/download/cp43-complete-issue/577 action=job_enqueue
08:27:19 processor.1 | [processor] [worker pool:aggrFooBar] 2019/04/16 08:27:19 perform: Starting download for Job{ID:aiKAl8wnyoZhBw, Aggr:aggrFooBar, URL:http://www.cartographicperspectives.org/index.php/journal/article/download/cp43-complete-issue/577, callback_url:, callback_type:kafka, callback_dst:dwl_images, UserAgent:} ...
08:27:29 processor.1 | [processor] [worker pool:aggrFooBar] 2019/04/16 08:27:29 perform: Download Failed for Job{ID:aiKAl8wnyoZhBw, Aggr:aggrFooBar, URL:http://www.cartographicperspectives.org/index.php/journal/article/download/cp43-complete-issue/577, callback_url:, callback_type:kafka, callback_dst:dwl_images, UserAgent:} Error while downloading file, net/http: request canceled (Client.Timeout exceeded while reading body), retriable: true, internal: false
```

Reply:
You are right :( is only respected if `download_timeout` is less than the client timeout:
```
curl -d '{"download_timeout": 1, "aggr_id":"aggrFooBar", "aggr_limit":8, "url":"http://www.cartographicperspectives.org/index.php/journal/article/download/cp43-complete-issue/577", "callback_type":"kafka" ,"callback_dst":"dwl_images", "extra":"foobar"}' http://localhost:8000/download
```

```
10:44:53 processor.1 | [processor] [worker pool:aggrFooBar] 2019/04/16 10:44:53 perform: Starting download for Job{ID:BgxAOZDXvDiI0Q, Aggr:aggrFooBar, URL:http://www.cartographicperspectives.org/index.php/journal/article/download/cp43-complete-issue/577, callback_url:, callback_type:kafka, callback_dst:dwl_images, UserAgent:} ...
10:44:54 processor.1 | [processor] [worker pool:aggrFooBar] 2019/04/16 10:44:54 perform: Download Failed for Job{ID:BgxAOZDXvDiI0Q, Aggr:aggrFooBar, URL:http://www.cartographicperspectives.org/index.php/journal/article/download/cp43-complete-issue/577, callback_url:, callback_type:kafka, callback_dst:dwl_images, UserAgent:} Error while downloading file, context deadline exceeded, retriable: true, internal: false
```

Reply:
So how are we going to handle this?

I see two options:

1) Create an HTTP client per Job instead of aggregation and set the timeout there
2) Add a HUGE default timeout.

Do you have any other (more elegant) ideas?

Reply:
So how are we going to handle this?

I see two options:

- Create an HTTP client per Job instead of aggregation and set the timeout there
- Add a HUGE default timeout.

Do you have any other (more elegant) ideas?

Reply:
(2) sounds fine to me. We could set it to 2 or even more hours.

Reply:
1. Adding a client per job, I think will be quite an overhead, it should be really necessary to do this IMHO.
2. We could do that,
3. Or, have also an other timeout per aggregation(client will be created with it) that will have the wanted scale that can be adjusted per job.

Reply:
Even if we go with the great sledgehammer (2) we should make the default timeout a configuration option to avoid having to re-build the project in order to change it!